datalad/metadata/metadata.py

```diff
@@ -264,163 +733,231 @@ def __call__(
+                # -> query through parent
+                # there must be a parent, otherwise this would be a non-dataset path
+                # and would have errored during annotation
+                to_query = ap['parentds']
+            else:
+                to_query = ap['path']
+            if to_query:
+                pcontent = content_by_ds.get(to_query, [])
+                pcontent.append(ap)
+                content_by_ds[to_query] = pcontent
 
-        content_by_ds, ds_props, completed, nondataset_paths = \
-            annotated2content_by_ds(
-                to_process,
-                refds_path=refds_path,
-                path_only=False)
-        assert(not completed)
+        # first deal with the two simple cases
+        if show_keys:
+            # report any dataset-defined keys and exit
+            for ds_path in content_by_ds:
+                db_path = opj(ds_path, db_relpath)
+                db = _load_json_object(db_path)
+                defs = db.get('definition', {})
+                for k in sorted(defs):
+                    ui.message('{}: {} ({}: {})'.format(
+                        ac.color_word(k, ac.BOLD),
+                        defs[k],
+                        ac.color_word('dataset', ac.MAGENTA),
+                        ds_path))
+            return
+        elif not (init or purge or reset or add or remove or define_key):
+            # just a query of metadata, no modification
+            # test for datasets that will be queried, but have never been aggregated
+            to_aggregate = [d for d in content_by_ds
+                            if not exists(opj(d, agginfo_relpath))]
+            if to_aggregate:
+                lgr.warning(
+                    'Metadata query results might be incomplete, initial '
+                    'metadata aggregation was not yet performed in datasets at: %s',
+                    to_aggregate)
 
+            for ds_path in content_by_ds:
+                DIRECT_QUERY = False
+                ds = Dataset(ds_path)
+                # sort requested paths into available components of this dataset
+                # and into things that might be available in aggregated metadata
+                query_agg = []
+                query_ds = []
+                for ap in content_by_ds[ds_path]:
+                    if ap.get('state', None) == 'absent':
+                        query_agg.append(ap)
+                    elif ap.get('type', None) == 'dataset' and not ap['path'] == ds_path:
+                        # this is an available subdataset, will be processed in another
+                        # iteration
+                        continue
+                    else:
+                        (query_ds if DIRECT_QUERY else query_agg).append(ap)
+                # report directly available metadata
+                if query_ds:
+                    pass
+                   # # TODO RF to use code form `aggregate`
```

Comment:
How is `pass` an alternative to using code from aggregate? ;-)

From reading I've no idea what is to be done here. If you're sure to not forget about this piece - I'm fine with it though.

Reply:
This code can never be reached. I am keeping it for now, until I know that the approach works in full.

Reply:
Ok.